### PR TITLE
Added Framework Runner and demo script

### PR DESCRIPTION
**Pull request by chelma**
### Description
* Added the core of the Framework's Runner, including logging, exception handling, state files, and the ability to resume a run from the state file
* Added some sample steps and a demo script which should be deleted in the next commit

### Issues Resolved
* https://github.com/opensearch-project/opensearch-migrations/issues/4

### Testing
* Added unit tests for the code
* Added demo script and steps, which are illustrated below:

Demoing a complete run
```
(.venv) chelma@3c22fba4e266 upgrades % ./run_demo.py
[FrameworkRunner] Running Step: Demo1
[Demo1] Checking if Docker is available...
[Demo1] Checking current running user...
[FrameworkRunner] Step Succeeded: Demo1
[FrameworkRunner] Running Step: Demo2
[Demo2] Pausing for 5 seconds...
[Demo2] You are: chelma
[FrameworkRunner] Step Succeeded: Demo2
[FrameworkRunner] Ran through all steps successfully
[FrameworkRunner] Saving application state to file...
[FrameworkRunner] Application state saved
[FrameworkRunner] Application state saved to: /tmp/utf/state-file
[FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2022-12-01_16_02_34
```

CTRL+C in the middle of the run to simulate a failure:
```
(.venv) chelma@3c22fba4e266 upgrades %
(.venv) chelma@3c22fba4e266 upgrades %
(.venv) chelma@3c22fba4e266 upgrades % ./run_demo.py
[FrameworkRunner] Running Step: Demo1
[Demo1] Checking if Docker is available...
[Demo1] Checking current running user...
[FrameworkRunner] Step Succeeded: Demo1
[FrameworkRunner] Running Step: Demo2
[Demo2] Pausing for 5 seconds...
^C[Demo2] User initiated a keyboard interrupt.  Aborting...
[Demo2] Keyboard interrupt
[FrameworkRunner] User aborted the operation
[FrameworkRunner] Saving application state to file...
[FrameworkRunner] Application state saved
[FrameworkRunner] Application state saved to: /tmp/utf/state-file
[FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2022-12-01_16_02_47
```

Demoing "resume" capability:
```
(.venv) chelma@3c22fba4e266 upgrades % ./run_demo.py --resume
[FrameworkRunner] Running Step: Demo2
[Demo2] Pausing for 5 seconds...
[Demo2] You are: chelma
[FrameworkRunner] Step Succeeded: Demo2
[FrameworkRunner] Ran through all steps successfully
[FrameworkRunner] Saving application state to file...
[FrameworkRunner] Application state saved
[FrameworkRunner] Application state saved to: /tmp/utf/state-file
[FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2022-12-01_16_02_56
```

Showing off what's in the statefile and log files:
```
(.venv) chelma@3c22fba4e266 upgrades % cat /tmp/utf/state-file
{
    "current_step": "Demo2",
    "exit_type": "Succeeded",
    "log_file": "/tmp/utf/logs/run.log.2022-12-01_16_02_56",
    "state_file": "/tmp/utf/state-file",
    "user": "chelma"
}
(.venv) chelma@3c22fba4e266 upgrades % cat /tmp/utf/logs/run.log.2022-12-01_16_02_56
2022-12-01 22:02:56.866433 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Running Step: Demo2
2022-12-01 22:02:56.866897 - upgrade_testing_framework.core.framework_step - [Demo2] Pausing for 5 seconds...
2022-12-01 22:03:01.868917 - upgrade_testing_framework.core.framework_step - [Demo2] You are: chelma
2022-12-01 22:03:01.869238 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Step Succeeded: Demo2
2022-12-01 22:03:01.869458 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Ran through all steps successfully
2022-12-01 22:03:01.869671 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Saving application state to file...
2022-12-01 22:03:01.870309 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Application state saved
2022-12-01 22:03:01.870433 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Application state saved to: /tmp/utf/state-file
2022-12-01 22:03:01.870542 - upgrade_testing_framework.core.framework_runner - [FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2022-12-01_16_02_56
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: upgrades/upgrade_testing_framework/core/logging_wrangler.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        return "[{}] {}".format(self.extra['step'], msg), kwargs
+
+class FrameworkLoggingFormatter(logging.Formatter):
+     def formatTime(self, record, datefmt=None):
+        return datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S.%f')
+
+LINE_SEP = '\u2063' # Invisible Unicode character.  Makes figuring out the beginning/end of a log entry much
+            # easier, given they can contain new line characters themselves.
+
+class LoggingWrangler:
```
Comment (lewijacn) [inline upgrades/upgrade_testing_framework/core/logging_wrangler.py:20]: Is this wrangler something we should be using for all our logging operations? Without this I imagine our DEBUG logs for those operations wouldn't appear in this timestamped logfile

File: upgrades/upgrade_testing_framework/core/framework_runner.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+import upgrade_testing_framework.core.constants as constants
+import upgrade_testing_framework.core.exception_base as exceptions
+from upgrade_testing_framework.core.framework_state import FrameworkState, get_initial_state
+from upgrade_testing_framework.core.framework_step import FrameworkStep
+from upgrade_testing_framework.core.logging_wrangler import FrameworkLoggingAdapter, LoggingWrangler
+from upgrade_testing_framework.core.workspace_wrangler import WorkspaceWrangler
+import upgrade_testing_framework.steps as steps
+
+ISSUE_LINK = "https://github.com/opensearch-project/opensearch-migrations/issues/new/choose"
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply upgrades/upgrade_testing_framework/core/framework_runner.py:14]: I'd second this. If there was anyway we could get close to even a one click for users to post an issue populated with their error details would be very nice. Maybe even something like this: https://github.com/sindresorhus/new-github-issue-url

**Review by lewijacn — APPROVED** [review]
Looks good, left some thoughts and questions I had

File: upgrades/upgrade_testing_framework/workflows/__init__.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from upgrade_testing_framework.steps import *
```
Comment (okhasawn) [inline upgrades/upgrade_testing_framework/workflows/__init__.py:1]: Chris said that "*" imports are evil and he should change this.

File: upgrades/setup.py [hunk 4]
```diff
@@ -8,7 +8,9 @@
     package_dir={"": "upgrade_testing_framework"},
     packages=setuptools.find_packages(where="upgrade_testing_framework"),
     install_requires=[
+        "coloredlogs",
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/setup.py:11]: I'm pretty it pulls the latest available from pypi when you do your python install command inside your virtual environment.  See the README for a bit more context.

File: upgrades/upgrade_testing_framework/core/workspace_wrangler.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+class WorkspaceWrangler:
+    """
+    This class is used to set up and track the directory tree that stores the framework's logs, state file, and any
+    other resources on disk.
+    """
+    def __init__(self, base_directory=DEFAULT_BASE_DIR):
+        self.base_directory = base_directory
+        self._ensure_workspace_exists()
+
+    def _ensure_workspace_exists(self):
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/core/workspace_wrangler.py:14]: The logs written to file are uncolored on purpose (ANSI codes are not good to log, IMO).  For terminal, the log line color depends on the log level.  INFO is standard text color, WARN is yellow, ERROR is red, etc.

File: upgrades/upgrade_testing_framework/core/logging_wrangler.py (hunk 1 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/core/logging_wrangler.py:20]: Yeah, it should be used everywhere, but you only need to set it up once in a given Python application.  Check the Python library docs on logging to get a better feel for how it works.  It's a bit complicated.

File: upgrades/upgrade_testing_framework/workflows/__init__.py (hunk 3 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/workflows/__init__.py:1]: Thanks Omar!  I'll fix it in the next commit.

**Pull request merged**